File: csharp/go-counting/GoCounting.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum Owner { None, Black, White }

public class GoCounting
{
    private string[] _board;

    public GoCounting(string input) => _board = input.Split('\n');

    public (Owner, HashSet<(int, int)>) Territory((int, int) c)
    {
        if (!OnBoard(c)) throw new ArgumentException("out of range");
        if (OwnerOf(c) != Owner.None) return (Owner.None, new());

        var (fields, edges) = (new HashSet<(int, int)> { c }, new HashSet<(int, int)>());
        var neighbors = Neighbors(c).ToArray();

        while (neighbors.Any())
        {
            foreach (var n in neighbors)
            {
                if (OwnerOf(n) == Owner.None) fields.Add(n);
                else edges.Add(n);
            }

            neighbors = neighbors
                .Where(n => OwnerOf(n) == Owner.None)
                .SelectMany(Neighbors)
                .Where(n => !fields.Contains(n) && !edges.Contains(n)).ToArray();
        }

        if (edges.Select(OwnerOf).Distinct().Count() == 1) 
             return (OwnerOf(edges.First()), fields);

        else return (Owner.None, fields);
    }

    public Dictionary<Owner, HashSet<(int, int)>> Territories()
    {
        var res = new Dictionary<Owner, HashSet<(int, int)>>
        {
            [Owner.None] = new(), [Owner.White] = new(), [Owner.Black] = new()
        };
        var visited = new bool[_board[0].Length, _board.Length];

        for (int x = 0; x < _board[0].Length; x++)
            for (int y = 0; y < _board.Length; y++)      
                if (!visited[x, y])
                {
                    visited[x, y] = true;
                    var (owner, territory) = Territory((x, y));

                    foreach (var (cx, cy) in territory)
                    {
                        res[owner].Add((cx, cy));
                        visited[cx, cy] = true;
                    }
                }  

        return res;
    }

    private (int, int)[] DELTAS = new[] { (-1, 0), (1, 0), (0, -1), (0, 1) };

    private IEnumerable<(int, int)> Neighbors((int, int) c) => 
        DELTAS.Select(d => (c.Item1 + d.Item1, c.Item2 + d.Item2)).Where(OnBoard);

    private bool OnBoard((int x, int y) c) => 
        c.x >= 0 && c.y >= 0 && c.x < _board[0].Length && c.y < _board.Length;

    private Owner OwnerOf((int x, int y) c) =>  _board[c.y][c.x] == 'W' ? Owner.White :
        _board[c.y][c.x] == 'B' ? Owner.Black : Owner.None;
}

File: fsharp/go-counting/GoCounting.fs
module GoCounting

type Owner = White | Black | None

type private Board = string list

let private width (board: Board) = board[0] |> String.length

let private height (board: Board) = board |> List.length

let private onBoard (board: Board) (x, y) = 
    (x >= 0) && (x < width board) && (y >= 0) && (y < height board)

let private neighbors (board: Board) (x, y) = 
    [| (-1, 0); (1, 0); (0, -1); (0, 1) |]
    |> Array.map (fun (dx, dy) -> (x + dx, y + dy)) 
    |> Array.filter (onBoard board)

let private ownerOf (board: Board) (x, y) = 
    match board[y][x] with 
    | 'W' -> White 
    | 'B' -> Black
    | _   -> None

let private emptyFields (board: Board) = 
    [for x in 0..(width board - 1) do 
        for y in 0..(height board - 1) do 
            if ownerOf board (x, y) = None then yield (x, y)]

let public territory board position : (Owner * (int * int) list) option = 
    if not (onBoard board position) then 
        Option.None 
    elif ownerOf board position <> None then 
        Some (None, [])
    else 
        let mutable fields = [position] |> Set.ofList
        let mutable edges = [] |> Set.ofList
        let ngbs = neighbors board position 
        let updateFieldsAndEdges p =
            match ownerOf board p with 
            | None -> fields <- fields |> Set.add p 
            | _    -> edges  <- edges  |> Set.add p
        let rec loop actualNgbs =
            if actualNgbs |> Array.isEmpty then ()
            else
                actualNgbs |> Array.map updateFieldsAndEdges |> ignore 
                loop (actualNgbs // get new neighbors and recurse
                |> Array.filter (fun p -> ownerOf board p = None)
                |> Array.collect (neighbors board)
                |> Array.filter (fun p -> not (Set.contains p fields || Set.contains p edges)))
        loop ngbs |> ignore
        let edgeOwners = edges |> Seq.map (ownerOf board) |> Seq.distinct
        if Seq.length edgeOwners = 1 then Some (Seq.head edgeOwners, Set.toList fields)
        else  Some (None, Set.toList fields) 

let public territories board : Map<Owner, (int * int) list> = 
    let resultMap = [(Black, []); (White, []); (None, [])] |> Map.ofList
    let empties = emptyFields board |> Set.ofList
    let rec loop emptyRest acc = 
        if emptyRest |> Set.isEmpty then acc
        else 
            let pos = emptyRest |> Seq.head 
            match territory board pos with
            | Some (owner, fields) -> 
                let newEntry =  Map.find owner acc |> List.append fields |> List.sort
                loop (Set.ofList fields |> Set.difference emptyRest) (Map.add owner newEntry acc)
            | _                    -> failwith "unreachable"
    loop empties resultMap